server_fn_macro/src/lib.rs

```diff
@@ -239,180 +248,572 @@ pub fn server_macro_impl(
     let func = if cfg!(feature = "ssr") {
         quote! {
             #docs
             #(#attrs)*
             #vis async fn #fn_name(#(#fn_args),*) #output_arrow #return_ty {
-                #block
+                #dummy_name(#(#field_names),*).await
             }
         }
     } else {
         quote! {
             #docs
             #(#attrs)*
             #[allow(unused_variables)]
-            #vis async fn #fn_name(#(#fn_args_2),*) #output_arrow #return_ty {
-                #server_fn_path::call_server_fn(
-                    &{
-                        let prefix = #struct_name::PREFIX.to_string();
-                        prefix + "/" + #struct_name::URL
-                    },
-                    #struct_name { #(#field_names_5),* },
-                    #encoding
-                ).await
+            #vis async fn #fn_name(#(#fn_args),*) #output_arrow #return_ty {
+                use #server_fn_path::ServerFn;
+                let data = #struct_name { #(#field_names),* };
+                data.run_on_client().await
             }
         }
     };
 
+    enum PathInfo {
+        Serde,
+        Rkyv,
+        None,
+    }
+
+    let (path, derives) = match input_ident.as_deref() {
+        Some("Rkyv") => (
+            PathInfo::Rkyv,
+            quote! {
+                Clone, #server_fn_path::rkyv::Archive, #server_fn_path::rkyv::Serialize, #server_fn_path::rkyv::Deserialize
+            },
+        ),
+        Some("MultipartFormData")
+        | Some("Streaming")
+        | Some("StreamingText") => (PathInfo::None, quote! {}),
+        Some("SerdeLite") => (
+            PathInfo::Serde,
+            quote! {
+                Clone, #server_fn_path::serde_lite::Serialize, #server_fn_path::serde_lite::Deserialize
+            },
+        ),
+        _ => (
+            PathInfo::Serde,
+            quote! {
+                Clone, #server_fn_path::serde::Serialize, #server_fn_path::serde::Deserialize
+            },
+        ),
+    };
+    let addl_path = match path {
+        PathInfo::Serde => quote! {
+            #[serde(crate = #serde_path)]
+        },
+        PathInfo::Rkyv => {
+            let rkyv_path = format!("{server_fn_path}::rkyv");
+            quote! {
+                #[archive(crate = #rkyv_path, check_bytes)]
+            }
+        }
+        PathInfo::None => quote! {},
+    };
+
+    // TODO reqwest
+    let client = quote! {
+        #server_fn_path::client::browser::BrowserClient
+    };
+
+    let req = if !cfg!(feature = "ssr") {
+        quote! {
+            #server_fn_path::request::BrowserMockReq
+        }
+    } else if cfg!(feature = "axum") {
+        quote! {
+            #server_fn_path::axum_export::http::Request<#server_fn_path::axum_export::body::Body>
+        }
+    } else if cfg!(feature = "actix") {
+        quote! {
+            #server_fn_path::request::actix::ActixRequest
+        }
+    } else {
+        return Err(syn::Error::new(
```

Comment:
How to ignore this error for other frameworks? Likes viz.

Reply:
<img width="861" alt="Screenshot 2024-01-20 at 03 54 41" src="https://github.com/leptos-rs/leptos/assets/27926/83e7033d-cae2-4e32-a078-a6ac718c7eda">

Reply:
This is a good point and I did not have a particularly good answer. Here are two proposals:
1. I can add `preset_req` and `preset_res` arguments to the `server_macro_impl` function, which `leptos_macro::server` will just pass `None`. Then you can create your own Viz `#[server]` macro, which would have a really simple implementation:
```rust
#[proc_macro_attribute]
#[proc_macro_error]
pub fn server(args: proc_macro::TokenStream, s: TokenStream) -> TokenStream {
    match server_fn_macro::server_macro_impl(
        args.into(),
        s.into(),
        Some(syn::parse_quote!(::leptos::server_fn)),
        "/api",
        // change these two to Some(viz::whatever::type)
        None,
        None,
    ) {
        Err(e) => e.to_compile_error().into(),
        Ok(s) => s.to_token_stream().into(),
    }
}

```
2.  I've added `req = ` and `res = ` arguments to the `#[server]` macro that allowed this to be specified, so
```rust
#[server(
  input = Json,
  req = SomeRequestType<Body>,
  res = SomeResponseType<Body>
)]
```
Doing this on every server function seems like a pain, so I suspect option 1 is better.

Reply:
I found both `Req` and `Res` trails should be implemented.